[osdp_fixes 4/5] Secure channel fixes

sidcha opened the pull request:
Some minor fixes to SC:

- [mgmt/osdp: sc: Restructure loop to avoid a bunch of -1s](https://github.com/zephyrproject-rtos/zephyr/commit/ceb1ceed685387021a5344a09ca8da43ffc97667) 
- [mgmt/osdp: Rework secure channel key management](https://github.com/zephyrproject-rtos/zephyr/commit/90f8e19d21e4cfa3b0876dcced13230671018f91) 
- [mgmt/osdp: phy: Allow non-conformant, 0 length, encrypted data blocks](https://github.com/zephyrproject-rtos/zephyr/commit/91252e5daeca381d4f829a47ad7afbc799ace841) 
- [mgmt/osdp: discard secure channel if KEYSET is ACKed in plaintext](https://github.com/zephyrproject-rtos/zephyr/commit/ccf2e16bb2f9c8b4d6d609ebc4ac1d0245fb01eb)

sidcha committed:
mgmt/osdp: sc: Restructure loop to avoid a bunch of -1s

Hoisting a `length--` helps simplify a lot of code which has ad-hoc -1s
in a  bunch of places. Also, fix some formatting issues and remove
unnecessary log lines.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Rework secure channel key management

OSDP command KEYSET is used to set the secure channel base key for all
connected PDs. This key is then used to derive the session keys for each
secure channel session. When the app wants to set the this key, it has
to issue a command and then both the CP/PD has to be notified of this
change so they can store this key in a non-volatile medium for future
operations across power cycles.

The current implementation of OSDP had limited support for key
management. This patch adds all the bells and whistles needed to handle
keyset command/event in the CP/PD application.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: phy: Allow non-conformant, 0 length, encrypted data blocks

If command or reply has no data, PD "should" use secure message types
SCS_15 or SCS_16. But some PD seem to not implement this correctly. We
will be tolerant towards those faulty implementations.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: discard secure channel if KEYSET is ACKed in plaintext

HID readers are responding to a KEYSET command with an ACK in plaintext
instead of using the current session keys to acknowledge this command
(which is the reasonable thing to do as the command itself was received
encrypted with the old key). Since the spec doesn't say anything about
this, both methods are technically correct.

Make changes to CP so it allows ACKs in plaintext for KEYSET command in
particular and make the PD implementation do this too as it makes the code
look cleaner there (perhaps HID did it for the same reasons :D).

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

r2r0 commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -241,10 +241,22 @@ static int cp_build_command(struct osdp_pd *pd, uint8_t *buf, int max_len)
 		buf[len++] = 1;  /* key type (1: SCBK) */
 		buf[len++] = 16; /* key length in bytes */
-		osdp_compute_scbk(pd, buf + len);
+		if (cmd->keyset.type == 1) { /* SCBK */
```
Consider adding and using some symbolic names (define/enum) for key types - it is better to rely on code not on comment.

sidcha replied:
Ack. Added another patch for this.

sidcha replied:
On second thought, I will not push it to this PR because it introduces an API change (that delays this PR from being merged). The next and final PR in this series does all the API changes, so I will queue this one there.

sidcha committed:
mgmt/osdp: Make sure we decrypt atleast one block

OSDP secure channel message blocks are terminated with a 0x80. This means
that even for zero length messages, we would have at least one block of
encrypted data to decrypt (since message blocks are rounded up to the next
16 byte boundary). The current length assertion checks for 16 byte
alignment but a malicious peer could send a specially crafted packet with
zero length blocks. Fix this issue by adding check for length == 0 case.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha commented on subsys/mgmt/osdp/src/osdp_sc.c:
```diff
@@ -154,15 +156,16 @@ int osdp_decrypt_data(struct osdp_pd *pd, int is_cmd, uint8_t *data, int length)
 	osdp_decrypt(pd->sc.s_enc, iv, data, length);
 
-	while (data[length - 1] == 0x00) {
+	length--;
```
Not it is not, good catch. It is not possible for a complying PD/CP to send a packet that would reach here with length <= 0 but a malicious PD can specially craft a packet to do that. I added commit to address this issue.

sidcha commented on subsys/mgmt/osdp/src/osdp_sc.c:
```diff
@@ -222,16 +221,17 @@ int osdp_compute_mac(struct osdp_pd *pd, int is_cmd,
-	if (ISSET_FLAG(pd, PD_FLAG_PD_MODE)) {
-		memcpy(key, pd->sc.scbk, 16);
+	if (preserve_scbk) {
+		memcpy(scbk, pd->sc.scbk, 16);
```
I agree with your point.. but it feels like that's not the kind of threat model we work with here(?). Anyhow, I cannot find a secure variant of of memset in zephyr with my grep-ing so I wonder what the others are doing about this..

sidcha replied:
@carlescufi @gmarull @erwango thoughts?

r2r0 replied:
> but it feels like that's not the kind of threat model we work with here(?). Anyhow, I cannot find a secure variant of of memset in zephyr with my grep-ing so I wonder what the others are doing about this..

Usually this is handled by crypto library (i.e. mbedTLS clears allocated memory before feeeing it), so in Zephyr code this kind of functionality is not usual.
In this case threat comes not from hostile memory access i.e. from another thread but from accidential disclosure of memory content (i.e. by dump logging in non related thread), so witch such assumption regular memset is good enough.

sidcha replied:
> so in Zephyr code this kind of functionality is not usual.

Possible. But I'd expect there to be other modules/subsystem where we invoke low level crypto APIs. To be fair, I haven't spent time looking for it seriously I just grep-ed for varying call sites for memset such as `*_memset_*`. I'll allocate some time for this; if I cannot find one, I'll add it.

> So witch such assumption regular memset is good enough.

No, regular memset gets optimized away by compilers with aggressive optimization levels/options.

r2r0 replied:
So we have options:
- ignore problem (leaving master password in memory is for our product a no go)
- add something like osdp_memset_s function (i.e. macro which will use standard memset_s in case of C11 compilation or something like in mbedtls_platform_zeroize otherwise)
- use crypto API which will allow to avoid using plain passwords for secure session setup (a little different topic than using plain password for KEYSET command)

sidcha committed:
mgmt/osdp: Clear secure channel base key from stack

Secure Channel Base Key (SCBK) is a secret key used to derive the session
keys used to encrypt and decrypt OSDP packets. Secure coding practice
requires us to clear such sensitive data from stack once we are done
needing it. This patch addresses this issue.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>
